src/settings.c

```diff
@@ -176,29 +176,52 @@ settings_create (GDBusConnection *connection,
 {
   Settings *settings;
   g_autoptr(GError) error = NULL;
-  int i;
+  int i, j;
+  int n_impls_tmp;
+  XdpImplSettings **impls_tmp;
 
-  n_impls = implementations->len;
-  impls = g_new (XdpImplSettings *, n_impls);
+  n_impls_tmp = implementations->len;
+  impls_tmp = g_new (XdpImplSettings *, n_impls_tmp);
 
   settings = g_object_new (settings_get_type (), NULL);
 
-  for (i = 0; i < n_impls; i++)
+  for (i = 0; i < n_impls_tmp; i++)
     {
       PortalImplementation *impl = g_ptr_array_index (implementations, i);
       const char *dbus_name = impl->dbus_name;
 
-      impls[i] = xdp_impl_settings_proxy_new_sync (connection,
-                                                   G_DBUS_PROXY_FLAGS_NONE,
-                                                   dbus_name,
-                                                   DESKTOP_PORTAL_OBJECT_PATH,
-                                                   NULL,
-                                                   &error);
-      if (impls[i] == NULL)
-        g_warning ("Failed to create settings proxy: %s", error->message);
+      impls_tmp[i] = xdp_impl_settings_proxy_new_sync (connection,
+                                                       G_DBUS_PROXY_FLAGS_NONE,
+                                                       dbus_name,
+                                                       DESKTOP_PORTAL_OBJECT_PATH,
+                                                       NULL,
+                                                       &error);
+      if (impls_tmp[i] == NULL)
+        {
+          g_warning ("Failed to create settings proxy: %s", error->message);
+        }
       else
-        g_signal_connect (impls[i], "setting-changed", G_CALLBACK (on_impl_settings_changed), settings);
+        {
+          n_impls++;
+          g_signal_connect (impls_tmp[i], "setting-changed", G_CALLBACK (on_impl_settings_changed), settings);
+        }
+    }
+
+  if (!n_impls)
+    {
+      g_free(impls_tmp);
+      return NULL;
+    }
+
+  impls = g_new (XdpImplSettings *, n_impls);
+
+  for (i = 0, j = 0; i < n_impls_tmp && j < n_impls; i++, j++)
+    {
+      if (impls_tmp[i] != NULL)
+        impls[j] = impls_tmp[i];
     }
 
+  g_free(impls_tmp);
+
```

Comment:
I think a much simpler approach is sufficient here. No need to loop twice. Just allocate impls to the full length, fill it up with as many non-NULL objects as we get, and set n_impls accordingly. Nevermind that the array has some extra room in the end.


```
impls = g_new (XdpImplSettings, n);
for (i = 0; j = 0; i < n; i++)
  {
      impl = xdp_impl_settings_proxy_new_sync (...);
      if (impl)
        impls[j++] = impl;
  }
n_impls = j;
```

Reply:
Thanks for adding a test!